src/librustc_passes/ast_validation.rs

```diff
@@ -851,6 +857,29 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
                              "`async fn` is not permitted in the 2015 edition").emit();
         }
     }
+
+    fn visit_fn_decl(&mut self, fn_decl: &'a FnDecl) {
+        fn_decl
+            .inputs
+            .iter()
+            .flat_map(|i| i.attrs.as_ref())
+            .filter(|attr| !attr.check_name(sym::allow) && !attr.check_name(sym::cfg)
+                && !attr.check_name(sym::cfg_attr) && !attr.check_name(sym::deny)
+                && !attr.check_name(sym::forbid) && !attr.check_name(sym::warn))
```

Comment:
Looks like this check rejects non-builtin attributes, which it shouldn't do.

Also, this would probably look better as `[sym::allow, sym::cfg, ...].contains(attr.name_or_empty())` instead of multiple `check_name`s.

Reply:
Wouldn't it be better to deny non-builtin attributes as user attributes aren't currently supported? Allowing an attribute that compiles and does nothing (for now) can be misleading.

Reply:
>Allowing an attribute that compiles and does nothing

Does nothing in `rustc`.
It may do something in `rustfmt` or other code processing tool, that's kind of the point of non-builtin non-macro attributes.

Reply:
And yes, macro attributes will be interpreted as non-macros in this position, and that's a bug that I'm going to fix (in June, tentatively), but it's ok, since they will produce a feature gate or other error anyway.

Reply:
Oh, sorry for the misinformation.
I had trouble with `is_builtin_attr` because it didn't take into account certain parameter like `#[test]` that aren't used by `BUILTIN_ATTRIBUTES`. Any clues?

Reply:
`#[test]` is a macro, not a built-in attribute.
What happens if you write `fn f(#[test] a: u8)`?
The compiler should report an error, if I'm not mistaken, something like ``attribute `test` is currently unknown to the compiler``.

Reply:
Aha, `visit.rs` forgets to visit the new attributes, that's why the error is not reported.

Reply:
Thanks! I will look into it

Reply:
@petrochenkov By visiting attributes on `fn_decl`, the compiler successfully denies `#[test]` but it is very awkward and could be misleading to users: `The attribute ``test`` is currently unknown to the compiler and may have meaning added to it in the future [E0658]`.
Wouldn't it be better to simply deny non-builtin attributes and macros?

Reply:
Thanks! This was my last concern and I will soon update the PR